tests/functional/curate/cram/validate-records.t

```diff
@@ -0,0 +1,37 @@
+Setup
+
+  $ source "$TESTDIR"/_setup.sh
+
+Testing records are validated appropriately by augur curate.
+Create NDJSON file for testing validation catches records with different fields.
+
+  $ cat >records.ndjson <<~~
+  > {"string": "string_1"}
+  > {"string": "string_2"}
+  > {"string": "string_3"}
+  > {"string": "string_4", "number": 123}
+  > ~~
+
+This will always pass thru the records that pass validation but should raise an
+error when it encounters the record with mismatched fields.
+
+  $ cat records.ndjson | ${AUGUR} curate passthru
+  ERROR: Records do not have the same fields! Please check your input data has the same fields.
+  {"string": "string_1"}
+  {"string": "string_2"}
+  {"string": "string_3"}
+  [2]
+
+Passing the records through multiple augur curate commands should raise the
+same error when it encounters the record with mismatched fields.
```

Comment:
[not a request for changes]

Do you have any good resources for understanding how pipes work in situations like this? I.e. the `augur curate X` is printing lines individually, so does an individual line flow through all `curate commands` before the next one starts flowing through? That's what this output makes it seem like. But when python flushes the print buffer must come into the equation right? And unix pipes presumably have some concept of backpressure / buffering?

Reply:
From my understanding, the shell starts all the commands at the same time, with file descriptors arranged so that the STDOUT of the first command is the STDIN of the second, and so on down the line. 

There is a buffer for the pipe, managed by the kernel, but if it's full, it blocks on the write side (and if it's empty, it blocks on the read side). [This SO answer may be helpful.](https://stackoverflow.com/questions/51452321/how-does-buffer-in-bash-pipe-work-on-linux#51452413)

Reply:
My understanding is as @genehack described ☝️ (I've only skimmed the [pipe man page](https://man7.org/linux/man-pages/man7/pipe.7.html) and have a limited understanding here)

> so does an individual NDJSON line flow through all curate commands before the next one starts flowing through?

This should depend on the buffer size, where multiple records can flow through a single command to fill up the buffer before being passed to the next command. 

In the case where the first command runs into an error, it should close it's "write end" so the subsequent commands will receive some end-of-file signal and terminate after writing their outputs as well. (Or exit immediately if `set -eo pipefail`)

Reply:
Thanks both! Reading those resources this is my understanding of what's happening (using `c1` for the first curate command, etc):
1. `c1` writes the first record to the appropriate fd, and there's no buffering on the python side (since we are using `print()` with the default new line ending).
1. `c2` reads this more-or-less immediately and in parallel with `c1` continuing to run. It writes output to its fd, `c3` reads it and so on.
1. This happens for the first three (valid) records - they make it through all three curate commands and are written to stdout. 
1. At some point `c1` reads the invalid record, prints to stderr, and exits code 2. This is, AFAICT, seen by `c2` no differently to an "end of file" and so `c2` exits (code 0) once it's consumed all the data in its input buffer.
1. The `pipefail` causes the entire pipeline to have exit code 2 because `c1` had exit code 2, but this is done after the pipeline has finished. I.e. doesn't actually change any behaviour of the pipeline --`c2` continues to run if it has data to read on it's input buffer and so on.
1. The order of steps (3) and (4) seems like it's a race-condition, but the fact that the pipeline's output has the error message (stderr) before the records (stdout) indicates that (4) comes first. Stdout seems to be line-buffered so I don't think that's important here.

Reply:
James asked me to check his understanding, and generally yeah, it's about right AFAICT.

One small inaccuracy is that `sys.stdout` is [block-buffered](https://docs.python.org/3/library/sys.html#sys.stdout), not line-buffered, when not connected to a TTY. (And it can be unbuffered entirely if desired.)

If you wanted to empirically test this understanding, you could `strace` the pipeline (e.g. the `bash` process and its children) and inspect the order of read/write/exit operations for each process in the pipeline. (`strace` is a Linux tool; there are equivalents for macOS, but I'm not as adept with them.)

Reply:
Additional things to consider if you want to dig into the order of 3 vs. 4 more is also how exactly stdout and stderr are interleaved by cram to match against the test file and the buffering mode of Python's `sys.stderr` (which is version dependent).